File: app/src/main/java/es/santirivera/surveilfall/di/modules/WebServicesModule.java
package es.santirivera.surveilfall.di.modules;

import android.content.Context;

import com.google.gson.Gson;

import java.io.IOException;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import es.santirivera.surveilfall.data.net.NetworkManager;
import es.santirivera.surveilfall.data.net.NetworkManagerImpl;
import es.santirivera.surveilfall.data.net.WServices;
import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class WebServicesModule {

    @Provides
    @Singleton
    public WServices provideWServices(Retrofit retrofit) {
        return retrofit.create(WServices.class);
    }

    @Provides
    @Singleton
    public Retrofit provideRetrofit(Gson gson, final Context context, final NetworkManager manager) {
        Cache cache = new Cache(context.getCacheDir(), 50 * 1024 * 1024);
        OkHttpClient okHttpClient = new OkHttpClient.Builder().cache(cache).addInterceptor(new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();
                if (manager.isInternetAvailable(context)) {
                    request.newBuilder().header("Cache-Control", "public, max-age=" + 5).build();
                } else {
                    request.newBuilder().header("Cache-Control", "public, only-if-cached, max-stale=" + 60 * 60 * 24 * 7).build();
                }
                return chain.proceed(request);
            }
        }).build();
        return new Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(okHttpClient)
                .baseUrl("https://api.scryfall.com")
                .build();
    }

    @Singleton
    @Provides
    NetworkManager providesNetworkManager(Context context) {
        return new NetworkManagerImpl(context);
    }
}

File: app/src/main/java/es/santirivera/surveilfall/di/modules/WebServicesModule.kt
package es.santirivera.surveilfall.di.modules

import android.content.Context

import com.google.gson.Gson

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import es.santirivera.surveilfall.data.net.NetworkManager
import es.santirivera.surveilfall.data.net.NetworkManagerImpl
import es.santirivera.surveilfall.data.net.ScryfallWebServices
import es.santirivera.surveilfall.data.net.GithubWebServices
import okhttp3.Cache
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

@Module
class WebServicesModule {

    @Provides
    @Singleton
    fun provideScryfallWebServices(gson: Gson, context: Context, manager: NetworkManager): ScryfallWebServices {
        val cache = Cache(context.cacheDir, (50 * 1024 * 1024).toLong())
        val okHttpClient = OkHttpClient.Builder().cache(cache).addInterceptor { chain ->
            val request = chain.request()
            if (manager.isInternetAvailable(context)) {
                request.newBuilder().header("Cache-Control", "public, max-age=" + 5).build()
            } else {
                request.newBuilder().header("Cache-Control", "public, only-if-cached, max-stale=" + 60 * 60 * 24 * 7).build()
            }
            chain.proceed(request)
        }.build()
        return Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(okHttpClient)
                .baseUrl("https://api.scryfall.com")
                .build().create(ScryfallWebServices::class.java)
    }

    @Provides
    @Singleton
    fun provideGithubWebServices(gson: Gson): GithubWebServices {
        return Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create(gson))
                .baseUrl("https://raw.githubusercontent.com/")
                .build().create(GithubWebServices::class.java)
    }


    @Singleton
    @Provides
    internal fun providesNetworkManager(context: Context): NetworkManager {
        return NetworkManagerImpl(context)
    }
}